src/sage/graphs/matching_covered_graph.py

```diff
@@ -2003,6 +2004,351 @@ def has_perfect_matching(G, algorithm='Edmonds', solver=None, verbose=0,
+            ...
+            ValueError: the input graph is not bipartite
+        """
+        if not self.is_bipartite():
+            raise ValueError('the input graph is not bipartite')
+
+        if self.order() < 6:
+            return (True, None, None) if coNP_certificate else True
+
+        A, B = self.bipartite_sets()
+        matching = set(self.get_matching())
+        matching_neighbor = {x: y for u, v, *_ in self.get_matching() for x, y in [(u, v), (v, u)]}
+
+        for e in list(self.get_matching())[:]:
+            u, v, *_ = e
+
+            # Let G denote the undirected graph self, and
+            # let the graph H(e) := G — u — v
+            H = Graph(self, multiedges=False)
+            H.delete_vertices([u, v])
+
+            if not H.is_matching_covered(list(matching - set([e]))):
+                if not coNP_certificate:
+                    return False
+
+                # Construct the digraph D(e)(A ∪ B, F) defined as follows:
+                from sage.graphs.digraph import DiGraph
+                D = DiGraph()
+
+                # For each edge (a, b) in E(H(e)) ∩ M with a in A, b —> a in D(e).
+                # For each edge (a, b) in E(H(e)) with a in A, a —> b in D(e).
+                for a, b, *_ in H.edge_iterator():
+
+                    if a in B:
+                        a, b = b, a
+
+                    D.add_edge((a, b))
+                    if matching_neighbor[a] == b:
+                        D.add_edge((b, a))
+
+                # H(e) is matching covered iff D(e) is strongly connected.
+                # Check if D(e) is strongly connected using Kosaraju's algorithm
+                def dfs(v, visited, neighbor_iterator):
+                    stack = [v]  # a stack of vertices
+
+                    while stack:
+                        v = stack.pop()
+                        visited.add(v)
+
+                        for u in neighbor_iterator(v):
+                            if u not in visited:
+                                stack.append(u)
+
+                root = next(D.vertex_iterator())
+
+                visited_in = set()
+                dfs(root, visited_in, D.neighbor_in_iterator)
+
+                visited_out = set()
+                dfs(root, visited_out, D.neighbor_out_iterator)
+
+                # Note that by definition of D(e), it follows that C ⊆ E(H(e)) — M.
+                # Thus, C is a cut of H(e), which has a shore X such that every edge of C is
+                # incident with a vertex in X ∩ B.
+
+                # Moreover, M — e is a perfect matching of H(e), and thus, |X ∩ A| = |X ∩ B|
+                # Consequently, Y := X + v is a shore of a nontrivial tight cut T of G
+
+                if len(visited_in) != D.order():
+                    X = {w for w in D if w in visited_in}
+                else:
+                    X = {w for w in D if w in visited_out}
+
+                # Compute the directed cut C of D(e)
+                C = []
```

Comment:
This part must be rewritten. You don't need to build `C`. You just need to know if there is an edge satisfying `(a in X) ^ (b in X)`, and if so if `a in A`. This defines the `color_class` and so the vertex to add to `X`.

For instance, it could be something like:
```py
for a, b in H.edge_iterator(labels=False):
    if (a in X) ^ (b in X):
        x = a if a in A else b
        color_class = x not in X
        break

X.add(u if (not color_class and u in A) or (color_class and u in B) else v)
```